Sources-controller running error: panic: runtime error: invalid memory address

**Describe the bug**
The source controller keeps carshloopbackoff
```
kubectl get pod -n knative-eventing 
NAME                                   READY   STATUS             RESTARTS   AGE
eventing-controller-6b95c5fcfd-l86rq   1/1     Running            0          25h
eventing-webhook-7b9fffb486-8shhq      1/1     Running            0          11h
imc-controller-85554dc899-8x8hv        1/1     Running            0          25h
imc-dispatcher-5c7b5cbf7c-kdf42        1/1     Running            0          26h
sources-controller-66b6c89fdb-jtfbp   0/1     CrashLoopBackOff   8          19m

```

 **Expected behavior**
All pod are running normally

 **To Reproduce**
Apply the deployment yaml.

 **Knative release version**.
Knative 0.10

 **Additional context**
```
$ kubectl describe pod -n knative-eventing  sources-controller-66b6c89fdb-jtfbp 
Name:           sources-controller-66b6c89fdb-jtfbp
Namespace:      knative-eventing
Node:           192.168.199.192/192.168.199.192
Start Time:     Thu, 05 Dec 2019 21:13:29 +0800
Labels:         app=sources-controller
                eventing.knative.dev/release=v0.10.0
                pod-template-hash=66b6c89fdb
Annotations:    sidecar.istio.io/inject: false
Status:         Running
IP:             172.20.12.103
Controlled By:  ReplicaSet/sources-controller-66b6c89fdb
Containers:
  controller:
    Container ID:   docker://a44224ded499a1f243c5c5d3a38a5521b716724e0703b8d809218f3791d555b4
    Image:          gcr.azk8s.cn/knative-releases/knative.dev/eventing/cmd/sources_controller@sha256:27f0b025ea16907db2b52cb220894b91cac627c6fbff914fce90ae4d987f1b0b
    Image ID:       docker-pullable://gcr.azk8s.cn/knative-releases/knative.dev/eventing/cmd/sources_controller@sha256:27f0b025ea16907db2b52cb220894b91cac627c6fbff914fce90ae4d987f1b0b
    Port:           9090/TCP
    Host Port:      0/TCP
    State:          Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Thu, 05 Dec 2019 21:16:38 +0800
      Finished:     Thu, 05 Dec 2019 21:16:40 +0800
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Thu, 05 Dec 2019 21:15:06 +0800
      Finished:     Thu, 05 Dec 2019 21:15:08 +0800
    Ready:          False
    Restart Count:  5
    Requests:
      cpu:     100m
      memory:  100Mi
    Environment:
      SYSTEM_NAMESPACE:           knative-eventing (v1:metadata.namespace)
      CONFIG_LOGGING_NAME:        config-logging
      CONFIG_OBSERVABILITY_NAME:  config-observability
      METRICS_DOMAIN:             knative.dev/sources
      CRONJOB_RA_IMAGE:           gcr.azk8s.cn/knative-releases/knative.dev/eventing/cmd/cronjob_receive_adapter@sha256:bdf30b32d57e1536fca5e436ce9c074827afd5f6f5eceed4cc0a8ca97096cefa
      APISERVER_RA_IMAGE:         gcr.azk8s.cn/knative-releases/knative.dev/eventing/cmd/apiserver_receive_adapter@sha256:67300996636df61f7271cfaed335b57daaea9496acebb6f8db68b9f81931b2db
    Mounts:
      /etc/config-logging from config-logging (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from eventing-source-controller-token-9fr27 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  config-logging:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      config-logging
    Optional:  false
  eventing-source-controller-token-9fr27:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  eventing-source-controller-token-9fr27
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     <none>
Events:
  Type     Reason     Age                   From                      Message
  ----     ------     ----                  ----                      -------
  Normal   Scheduled  3m21s                 default-scheduler         Successfully assigned knative-eventing/sources-controller-66b6c89fdb-jtfbp to 192.168.199.192
  Normal   Pulled     104s (x5 over 3m20s)  kubelet, 192.168.199.192  Container image "gcr.azk8s.cn/knative-releases/knative.dev/eventing/cmd/sources_controller@sha256:27f0b025ea16907db2b52cb220894b91cac627c6fbff914fce90ae4d987f1b0b" already present on machine
  Normal   Created    104s (x5 over 3m20s)  kubelet, 192.168.199.192  Created container controller
  Normal   Started    104s (x5 over 3m20s)  kubelet, 192.168.199.192  Started container controller
  Warning  BackOff    77s (x10 over 3m16s)  kubelet, 192.168.199.192  Back-off restarting failed container

$  kubectl logs -n knative-eventing  sources-controller-66b6c89fdb-jtfbp
{"level":"info","ts":"2019-12-05T13:16:39.254Z","logger":"source_controller","caller":"sharedmain/main.go:165","msg":"Starting informers.","commit":"5fe2b0a"}
{"level":"info","ts":"2019-12-05T13:16:40.155Z","logger":"source_controller","caller":"sharedmain/main.go:171","msg":"Starting controllers...","commit":"5fe2b0a"}
{"level":"info","ts":"2019-12-05T13:16:40.156Z","logger":"source_controller.cronjob-source-controller","caller":"controller/controller.go:286","msg":"Starting controller and workers","commit":"5fe2b0a","knative.dev/controller":"cronjob-source-controller"}
{"level":"info","ts":"2019-12-05T13:16:40.156Z","logger":"source_controller.cronjob-source-controller","caller":"controller/controller.go:296","msg":"Started workers","commit":"5fe2b0a","knative.dev/controller":"cronjob-source-controller"}
{"level":"info","ts":"2019-12-05T13:16:40.156Z","logger":"source_controller.container-source-controller","caller":"controller/controller.go:286","msg":"Starting controller and workers","commit":"5fe2b0a","knative.dev/controller":"container-source-controller"}
{"level":"info","ts":"2019-12-05T13:16:40.156Z","logger":"source_controller.container-source-controller","caller":"controller/controller.go:296","msg":"Started workers","commit":"5fe2b0a","knative.dev/controller":"container-source-controller"}
{"level":"info","ts":"2019-12-05T13:16:40.156Z","logger":"source_controller.apiserver-source-controller","caller":"controller/controller.go:286","msg":"Starting controller and workers","commit":"5fe2b0a","knative.dev/controller":"apiserver-source-controller"}
{"level":"info","ts":"2019-12-05T13:16:40.156Z","logger":"source_controller.apiserver-source-controller","caller":"controller/controller.go:296","msg":"Started workers","commit":"5fe2b0a","knative.dev/controller":"apiserver-source-controller"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x14e3b35]

goroutine 152 [running]:
knative.dev/eventing/pkg/reconciler/containersource.(*Reconciler).setSinkURIArg(0xc000554c30, 0x1c964e0, 0xc0007d2cf0, 0xc000120700, 0xc000591920, 0x0, 0x19)
	/home/prow/go/src/knative.dev/eventing/pkg/reconciler/containersource/containersource.go:201 +0x245
knative.dev/eventing/pkg/reconciler/containersource.(*Reconciler).reconcile(0xc000554c30, 0x1c964e0, 0xc0007d2cf0, 0xc000120700, 0x0, 0x0)
	/home/prow/go/src/knative.dev/eventing/pkg/reconciler/containersource/containersource.go:148 +0x3da
knative.dev/eventing/pkg/reconciler/containersource.(*Reconciler).Reconcile(0xc000554c30, 0x1c964e0, 0xc0007d2cf0, 0xc00089f180, 0x19, 0xc0000ab7f8, 0x1c964e0)
	/home/prow/go/src/knative.dev/eventing/pkg/reconciler/containersource/containersource.go:94 +0x182
knative.dev/eventing/vendor/knative.dev/pkg/controller.(*Impl).processNextWorkItem(0xc000548f80, 0x1993a00)
	/home/prow/go/src/knative.dev/eventing/vendor/knative.dev/pkg/controller/controller.go:342 +0x76b
knative.dev/eventing/vendor/knative.dev/pkg/controller.(*Impl).Run.func2(0xc00096ed10, 0xc000548f80)
	/home/prow/go/src/knative.dev/eventing/vendor/knative.dev/pkg/controller/controller.go:291 +0x51
created by knative.dev/eventing/vendor/knative.dev/pkg/controller.(*Impl).Run
	/home/prow/go/src/knative.dev/eventing/vendor/knative.dev/pkg/controller/controller.go:289 +0x19e
```

--- a/pkg/reconciler/containersource/containersource.go
+++ b/pkg/reconciler/containersource/containersource.go
@@ -198,7 +198,7 @@ func (r *Reconciler) setSinkURIArg(ctx context.Context, source *v1alpha1.Contain
 
 	sinkURI, err := r.sinkResolver.URIFromDestination(*dest, source)
 	if err != nil {
-		source.Status.MarkNoSink("NotFound", `Couldn't get Sink URI from "%s/%s"`, dest.Ref.Namespace, dest.Ref.Name)
+		source.Status.MarkNoSink("NotFound", `Couldn't get Sink URI from %+v`, dest)
 		return fmt.Errorf("getting sink URI: %v", err)
 	}
 	if source.Spec.Sink.DeprecatedAPIVersion != "" &&

--- a/pkg/reconciler/containersource/containersource_test.go
+++ b/pkg/reconciler/containersource/containersource_test.go
@@ -68,6 +68,14 @@ var (
 			APIVersion: "eventing.knative.dev/v1alpha1",
 		},
 	}
+	nonsinkDestWithNamespace = duckv1beta1.Destination{
+		Ref: &corev1.ObjectReference{
+			Name:       sinkName,
+			Namespace:  testNS,
+			Kind:       "Trigger",
+			APIVersion: "eventing.knative.dev/v1alpha1",
+		},
+	}
 
 	deploymentName = fmt.Sprintf("containersource-%s-%s", sourceName, sourceUID)
 
@@ -83,9 +91,30 @@ var (
 	}
 	serviceURI = fmt.Sprintf("http://%s.%s.svc.cluster.local/", sinkName, testNS)
 
+	deprecatedSinkDest = duckv1beta1.Destination{
+		DeprecatedName:       sinkName,
+		DeprecatedKind:       "Channel",
+		DeprecatedAPIVersion: "messaging.knative.dev/v1alpha1",
+	}
+
+	deprecatedSinkDestWithNamespace = duckv1beta1.Destination{
+		DeprecatedName:       sinkName,
+		DeprecatedKind:       "Channel",
+		DeprecatedNamespace:  testNS,
+		DeprecatedAPIVersion: "messaging.knative.dev/v1alpha1",
+	}
+
 	sinkDest = duckv1beta1.Destination{
 		Ref: &corev1.ObjectReference{
 			Name:       sinkName,
+			Kind:       "Channel",
+			APIVersion: "messaging.knative.dev/v1alpha1",
+		},
+	}
+	sinkDestWithNamespace = duckv1beta1.Destination{
+		Ref: &corev1.ObjectReference{
+			Name:       sinkName,
+			Namespace:  testNS,
 			Kind:       "Channel",
 			APIVersion: "messaging.knative.dev/v1alpha1",
 		},
@@ -140,7 +169,7 @@ func TestAllCases(t *testing.T) {
 					// Status Update:
 					WithInitContainerSourceConditions,
 					WithContainerSourceStatusObservedGeneration(generation),
-					WithContainerSourceSinkNotFound(`Couldn't get Sink URI from "testnamespace/testsink"`),
+					WithContainerSourceSinkNotFound(fmt.Sprintf("Couldn't get Sink URI from %+v", &sinkDestWithNamespace)),
 				),
 			}},
 		}, {
@@ -170,7 +199,7 @@ func TestAllCases(t *testing.T) {
 					// Status Update:
 					WithInitContainerSourceConditions,
 					WithContainerSourceStatusObservedGeneration(generation),
-					WithContainerSourceSinkNotFound(`Couldn't get Sink URI from "testnamespace/testsink"`),
+					WithContainerSourceSinkNotFound(fmt.Sprintf("Couldn't get Sink URI from %+v", &nonsinkDestWithNamespace)),
 				),
 			}},
 		}, {
@@ -200,7 +229,37 @@ func TestAllCases(t *testing.T) {
 					// Status Update:
 					WithInitContainerSourceConditions,
 					WithContainerSourceStatusObservedGeneration(generation),
-					WithContainerSourceSinkNotFound(`Couldn't get Sink URI from "testnamespace/testsink"`),
+					WithContainerSourceSinkNotFound(fmt.Sprintf("Couldn't get Sink URI from %+v", &sinkDestWithNamespace)),
+				),
+			}},
+		}, {}, {
+			Name: "deprecated sink not ready",
+			Objects: []runtime.Object{
+				NewContainerSource(sourceName, testNS,
+					WithContainerSourceSpec(sourcesv1alpha1.ContainerSourceSpec{
+						DeprecatedImage: image,
+						Sink:            &deprecatedSinkDest,
+					}),
+					WithContainerSourceObjectMetaGeneration(generation),
+				),
+				NewChannel(sinkName, testNS),
+			},
+			Key:     testNS + "/" + sourceName,
+			WantErr: true,
+			WantEvents: []string{
+				Eventf(corev1.EventTypeWarning, "SetSinkURIFailed", `Failed to set Sink URI: getting sink URI: address not set for &ObjectReference{Kind:Channel,Namespace:testnamespace,Name:testsink,UID:,APIVersion:messaging.knative.dev/v1alpha1,ResourceVersion:,FieldPath:,}`),
+			},
+			WantStatusUpdates: []clientgotesting.UpdateActionImpl{{
+				Object: NewContainerSource(sourceName, testNS,
+					WithContainerSourceSpec(sourcesv1alpha1.ContainerSourceSpec{
+						DeprecatedImage: image,
+						Sink:            &deprecatedSinkDest,
+					}),
+					WithContainerSourceObjectMetaGeneration(generation),
+					// Status Update:
+					WithInitContainerSourceConditions,
+					WithContainerSourceStatusObservedGeneration(generation),
+					WithContainerSourceSinkNotFound(fmt.Sprintf("Couldn't get Sink URI from %+v", &deprecatedSinkDestWithNamespace)),
 				),
 			}},
 		}, {

